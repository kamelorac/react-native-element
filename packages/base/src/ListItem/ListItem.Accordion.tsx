import React from 'react';
import { Animated } from 'react-native';
import { ListItemBase, ListItemProps } from './ListItem';
import { ListItemContent } from './ListItem.Content';
import { Icon, IconNode } from '../Icon';
import { renderNode, RneFunctionComponent } from '../helpers';

export interface ListItemAccordionProps extends ListItemProps {
  /** Decide if Accordion is Expanded. */
  isExpanded?: boolean;

  /** Icon for unexpanded Accordion. */
  icon?: IconNode;

  /** Icon when Accordion is expanded, if not provided `icon` will be rotated 180deg. */
  expandIcon?: IconNode;

  /** Similar to ListItem's child. */
  content?: React.ReactNode;

  /** Don't rotate when Accordion is expanded. */
  noRotation?: boolean;

  /** Rotate Icon left side */
  leftRotate?: boolean;

  /** Don't show accordion icon. */
  noIcon?: boolean;

  /** Decide whether to show animation.
   * @default Object with duration 350ms and type timing
   * @type Animated.TimingAnimationConfig
   */
  animation?:
    | {
        type?: 'timing' | 'spring';
        duration?: number;
      }
    | boolean;
}

/** This allows making a accordion list which can show/hide content. */
export const ListItemAccordion: RneFunctionComponent<
  ListItemAccordionProps
> = ({
  children,
  isExpanded = false,
  icon = <Icon name={'chevron-down'} type="material-community" />,
  expandIcon,
  content,
  leftRotate = false,
  noRotation,
  noIcon,
  animation = {
    duration: 350,
    type: 'timing',
  },
  ...rest
}) => {
  const transition = React.useRef(new Animated.Value(0));

  const startAnimation = React.useCallback(() => {
    if (typeof animation !== 'boolean') {
      Animated[animation.type || 'timing'](transition.current, {
        toValue: Number(isExpanded),
        useNativeDriver: true,
        duration: animation.duration || 350,
      }).start();
    }
  }, [isExpanded, animation]);

  React.useEffect(() => {
    startAnimation();
  }, [isExpanded, startAnimation]);

  const iconAnimation = React.useMemo(
    () => ({
      transform: [
        {
          rotate: noRotation
            ? '0deg'
            : transition.current.interpolate({
                inputRange: [0, 1],
                outputRange: ['0deg', leftRotate ? '180deg' : '-180deg'],
              }),
        },
      ],
    }),
    [leftRotate, noRotation]
  );

  return (
    <>
      <ListItemBase {...rest}>
        {React.isValidElement(content) ? content : <ListItemContent />}
        {!noIcon && (
          <Animated.View
            testID="RNE__ListItem__Accordion__Icon"
            style={iconAnimation}
          >
            {renderNode(Icon, isExpanded ? expandIcon ?? icon : icon)}
          </Animated.View>
        )}
      </ListItemBase>
      {isExpanded && (
        <Animated.View
          testID="RNE__ListItem__Accordion__Children"
          style={{
            opacity: transition.current,
          }}
        >
          {children}
        </Animated.View>
      )}
    </>
  );
};

ListItemAccordion.displayName = 'ListItem.Accordion';
