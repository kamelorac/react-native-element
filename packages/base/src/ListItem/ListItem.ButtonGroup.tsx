import React from 'react';
import { StyleSheet } from 'react-native';
import { ButtonGroupProps, ButtonGroup } from '../ButtonGroup';
import { RneFunctionComponent } from '../helpers';

export interface ListItemButtonGroupProps extends ButtonGroupProps {}

/** This allows adding ButtonGroup to the ListItem.
 * This, Receives all [ButtonGroup](buttongroup#props) props. */
export const ListItemButtonGroup: RneFunctionComponent<
  ListItemButtonGroupProps
> = ({ containerStyle, ...rest }) => {
  return (
    <ButtonGroup
      {...rest}
      containerStyle={StyleSheet.flatten([styles.container, containerStyle])}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginHorizontal: 0,
    marginVertical: 0,
  },
});

ListItemButtonGroup.displayName = 'ListItem.ButtonGroup';
