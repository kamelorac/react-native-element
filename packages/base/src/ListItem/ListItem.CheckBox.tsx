import React from 'react';
import { StyleSheet } from 'react-native';
import { CheckBoxProps, CheckBox } from '../CheckBox';
import { RneFunctionComponent } from '../helpers';

export interface ListItemCheckBoxProps extends CheckBoxProps {}

/** This allows adding CheckBox to the ListItem.
 * This, Receives all [CheckBox](checkbox#props) props. */
export const ListItemCheckBox: RneFunctionComponent<ListItemCheckBoxProps> = ({
  containerStyle,
  ...rest
}) => {
  return (
    <CheckBox
      {...rest}
      containerStyle={StyleSheet.flatten([styles.container, containerStyle])}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    margin: 0,
    marginRight: 0,
    marginLeft: 0,
    padding: 0,
  },
});

ListItemCheckBox.displayName = 'ListItem.CheckBox';
