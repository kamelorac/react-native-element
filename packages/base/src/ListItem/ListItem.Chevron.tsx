import React from 'react';
import { StyleSheet, Platform } from 'react-native';
import { RneFunctionComponent } from '../helpers';
import { IconProps, Icon } from '../Icon';

export interface ListItemChevronProps extends Partial<IconProps> {}

/** This allows adding a Chevron Icon(arrow) to the ListItem.
 * This, Receives all [Icon](icon#props) props. */
export const ListItemChevron: RneFunctionComponent<ListItemChevronProps> = ({
  containerStyle,
  ...rest
}: Partial<IconProps>) => {
  return (
    <Icon
      type={Platform.OS === 'ios' ? 'ionicon' : 'material'}
      color="#D1D1D6"
      name={
        Platform.OS === 'ios'
          ? 'chevron-forward-outline'
          : 'keyboard-arrow-right'
      }
      size={16}
      containerStyle={StyleSheet.flatten([
        { alignSelf: 'center' },
        containerStyle,
      ])}
      {...rest}
    />
  );
};

ListItemChevron.displayName = 'ListItem.Chevron';
