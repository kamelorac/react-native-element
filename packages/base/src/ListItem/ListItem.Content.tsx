import React from 'react';
import { StyleSheet, View } from 'react-native';
import { RneFunctionComponent } from '../helpers';
import { TextProps } from '../Text';

export interface ListItemContentProps extends TextProps {
  right?: boolean;
}

/** This allows adding content to the ListItem.
 * This, Receives all [View](https://reactnative.dev/docs/view#props) props. */
export const ListItemContent: RneFunctionComponent<ListItemContentProps> = ({
  style,
  right,
  children,
  ...rest
}) => {
  const containerStyle = right ? styles.rightContainer : styles.container;
  return (
    <View style={[containerStyle, style]} {...rest}>
      {children}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  rightContainer: {
    flex: 0.5,
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
});

ListItemContent.displayName = 'ListItem.Content';
