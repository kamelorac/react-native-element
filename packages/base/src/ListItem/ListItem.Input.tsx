import React from 'react';
import { StyleSheet, TextInput } from 'react-native';
import { Theme } from '../helpers';
import { Input, InputProps } from '../Input';

export interface ListItemInputProps extends InputProps {}

/** This allows adding an Text Input within the ListItem.
 * This, Receives all [Input](Input.mdx#props) props. */
export const ListItemInput = React.forwardRef<
  TextInput,
  ListItemInputProps & { theme?: Theme }
>(({ inputStyle, inputContainerStyle, containerStyle, ...rest }, ref) => {
  return (
    <Input
      // @ts-ignore
      ref={ref}
      renderErrorMessage={false}
      {...rest}
      inputStyle={StyleSheet.flatten([styles.input, inputStyle])}
      inputContainerStyle={StyleSheet.flatten([
        styles.inputContainer,
        inputContainerStyle,
      ])}
      containerStyle={StyleSheet.flatten([styles.container, containerStyle])}
    />
  );
});

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingRight: 0,
  },
  inputContainer: {
    flex: 1,
    borderBottomWidth: 0,
    width: null,
    height: null,
  },
  input: {
    flex: 1,
    textAlign: 'right',
    width: null,
    height: null,
  },
});

ListItemInput.displayName = 'ListItem.Input';
