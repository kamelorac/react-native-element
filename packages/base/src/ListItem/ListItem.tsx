import React from 'react';
import {
  Platform,
  StyleProp,
  StyleSheet,
  Pressable,
  PressableProps,
  View,
  ViewStyle,
} from 'react-native';
import { PadView } from './components/PadView';
import { defaultTheme, RneFunctionComponent } from '../helpers';

export interface ListItemProps extends PressableProps {
  /** Additional main container styling. */
  containerStyle?: StyleProp<ViewStyle>;

  /** Specific styling to be used when list item is disabled. */
  disabledStyle?: StyleProp<ViewStyle>;

  /** Add divider at the top of the list item. */
  topDivider?: boolean;

  /** Add divider at the bottom of the list item. */
  bottomDivider?: boolean;

  /** Adds spacing between the leftComponent, the title component & right component. */
  pad?: number;

  /** Replace element with custom element. */
  Component?: typeof React.Component;

  /** Container for linear gradient. */
  ViewComponent?: typeof React.Component;

  /** Props for linear gradient component. */
  linearGradientProps?: any;

  /** Add enclosed children. */
  children?: any;
}

/** ListItems are used to display rows of information, such as a contact list, playlist, or menu.
 * They are very customizable and can contain switches, avatars, badges, icons, and more. */
export const ListItemBase: RneFunctionComponent<ListItemProps> = (props) => {
  const {
    containerStyle,
    onPress,
    onLongPress,
    Component = onPress || onLongPress ? Pressable : View,
    disabled,
    disabledStyle,
    bottomDivider,
    topDivider,
    pad = 16,
    linearGradientProps,
    ViewComponent = View,
    theme = defaultTheme,
    children,
    ...rest
  } = props;

  if (props.linearGradientProps && !props.ViewComponent) {
    console.warn(
      "You need to pass a ViewComponent to use linearGradientProps !\nExample: ViewComponent={require('react-native-linear-gradient')}"
    );
  }

  return (
    <Component
      {...rest}
      onPress={onPress}
      onLongPress={onLongPress}
      disabled={disabled}
    >
      <PadView
        Component={ViewComponent}
        {...linearGradientProps}
        style={StyleSheet.flatten([
          {
            ...Platform.select({
              ios: {
                padding: 14,
              },
              default: {
                padding: 16,
              },
            }),
            flexDirection: 'row',
            alignItems: 'center',
            backgroundColor: theme?.colors?.background,
            borderColor: theme?.colors?.divider,
          },
          topDivider && { borderTopWidth: StyleSheet.hairlineWidth },
          bottomDivider && { borderBottomWidth: StyleSheet.hairlineWidth },
          containerStyle,
          disabled && disabledStyle,
        ])}
        pad={pad}
      >
        {children}
      </PadView>
    </Component>
  );
};

ListItemBase.displayName = 'ListItem';
