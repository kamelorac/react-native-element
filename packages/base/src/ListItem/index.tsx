import { ListItemBase, ListItemProps } from './ListItem';
import { ListItemContent, ListItemContentProps } from './ListItem.Content';
import { ListItemChevron, ListItemChevronProps } from './ListItem.Chevron';
import { ListItemInput, ListItemInputProps } from './ListItem.Input';
import { ListItemCheckBox, ListItemCheckBoxProps } from './ListItem.CheckBox';
import {
  ListItemButtonGroup,
  ListItemButtonGroupProps,
} from './ListItem.ButtonGroup';
import { ListItemTitle, ListItemTitleProps } from './ListItem.Title';
import { ListItemSubtitle, ListItemSubtitleProps } from './ListItem.Subtitle';
import {
  ListItemSwipeable,
  ListItemSwipeableProps,
} from './ListItem.Swipeable';
import {
  ListItemAccordion,
  ListItemAccordionProps,
} from './ListItem.Accordion';

export const ListItem = Object.assign(ListItemBase, {
  Accordion: ListItemAccordion,
  ButtonGroup: ListItemButtonGroup,
  CheckBox: ListItemCheckBox,
  Chevron: ListItemChevron,
  Content: ListItemContent,
  Input: ListItemInput,
  Subtitle: ListItemSubtitle,
  Swipeable: ListItemSwipeable,
  Title: ListItemTitle,
});

export type {
  ListItemAccordionProps,
  ListItemButtonGroupProps,
  ListItemCheckBoxProps,
  ListItemChevronProps,
  ListItemContentProps,
  ListItemInputProps,
  ListItemProps,
  ListItemSubtitleProps,
  ListItemSwipeableProps,
  ListItemTitleProps,
};
