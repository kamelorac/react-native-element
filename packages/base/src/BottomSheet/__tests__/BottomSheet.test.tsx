import React from 'react';
import { BottomSheet } from '../index';
import { Modal, ScrollView } from 'react-native';
import { ListItem } from '../../ListItem/index';
import { fireEvent, renderWithWrapper } from '../../../.ci/testHelper';
import { SafeAreaView } from 'react-native-safe-area-context';

describe('BottomSheet Component', () => {
  it('renders correctly', () => {
    const list = [{ title: 'test' }, { title: 'test2' }];
    const tree = renderWithWrapper(
      <BottomSheet isVisible>
        {list.map((l, i) => (
          <ListItem key={i}>
            <ListItem.Content>
              <ListItem.Title>{l.title}</ListItem.Title>
            </ListItem.Content>
          </ListItem>
        ))}
      </BottomSheet>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should shows when isVisible is true', () => {
    const list = [{ title: 'test' }, { title: 'test2' }];
    const component = renderWithWrapper(
      <BottomSheet isVisible>
        {list.map((l, i) => (
          <ListItem key={i}>
            <ListItem.Content>
              <ListItem.Title>{l.title}</ListItem.Title>
            </ListItem.Content>
          </ListItem>
        ))}
      </BottomSheet>
    );
    expect(component.wrapper.findByType(Modal).props.visible).toBeTruthy();
  });

  it('should not show when isVisible is false', () => {
    const list = [{ title: 'test' }, { title: 'test2' }];
    const component = renderWithWrapper(
      <BottomSheet>
        {list.map((l, i) => (
          <ListItem key={i}>
            <ListItem.Content>
              <ListItem.Title>{l.title}</ListItem.Title>
            </ListItem.Content>
          </ListItem>
        ))}
      </BottomSheet>
    );
    expect(component.wrapper.findByType(Modal).props.visible).toBeFalsy();
  });

  it('should click the backdrop and use passed handler', () => {
    const onBackdropPress = jest.fn();
    const list = [{ title: 'test' }, { title: 'test2' }];
    const { wrapper } = renderWithWrapper(
      <BottomSheet isVisible onBackdropPress={onBackdropPress}>
        {list.map((l, i) => (
          <ListItem key={i}>
            <ListItem.Content>
              <ListItem.Title>{l.title}</ListItem.Title>
            </ListItem.Content>
          </ListItem>
        ))}
      </BottomSheet>,
      'RNE__Overlay__backdrop'
    );
    fireEvent.press(wrapper);
    expect(onBackdropPress).toHaveBeenCalled();
  });

  it('should render with the provided containerStyle', () => {
    const { wrapper } = renderWithWrapper(
      <BottomSheet
        isVisible={true}
        containerStyle={{ backgroundColor: 'rgba(1, 0.5, 0.25, 1.0)' }}
      />
    );
    expect(wrapper.findByType(SafeAreaView).props.style).toMatchObject({
      backgroundColor: 'rgba(1, 0.5, 0.25, 1.0)',
    });
  });

  it('should render ScrollView with the provided scrollViewProps', () => {
    const { wrapper } = renderWithWrapper(
      <BottomSheet
        isVisible={true}
        scrollViewProps={{ keyboardShouldPersistTaps: 'handled' }}
      />
    );
    expect(wrapper.findByType(ScrollView).props.keyboardShouldPersistTaps).toBe(
      'handled'
    );
  });
});
