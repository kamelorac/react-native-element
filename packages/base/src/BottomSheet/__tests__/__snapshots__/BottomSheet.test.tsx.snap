// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet Component renders correctly 1`] = `
<View
  testID="wrapper"
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          undefined,
        ]
      }
      testID="RNE__Overlay__backdrop"
    />
    <RNCSafeAreaView
      edges={
        [
          "bottom",
          "left",
          "right",
          "top",
        ]
      }
      pointerEvents="box-none"
      style={
        {
          "backgroundColor": "rgba(0,0,0,0.2)",
          "flex": 1,
          "flexDirection": "column-reverse",
        }
      }
    >
      <View>
        <RCTScrollView>
          <View>
            <View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#ffffff",
                    "borderColor": "#bcbbc1",
                    "flexDirection": "row",
                    "padding": 14,
                  }
                }
                testID="RNE__LISTITEM__padView"
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "flex-start",
                        "flex": 1,
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "backgroundColor": "transparent",
                        "color": "#242424",
                        "fontSize": 17,
                      }
                    }
                    testID="listItemTitle"
                  >
                    test
                  </Text>
                </View>
              </View>
            </View>
            <View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#ffffff",
                    "borderColor": "#bcbbc1",
                    "flexDirection": "row",
                    "padding": 14,
                  }
                }
                testID="RNE__LISTITEM__padView"
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "flex-start",
                        "flex": 1,
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    accessibilityRole="text"
                    style={
                      {
                        "backgroundColor": "transparent",
                        "color": "#242424",
                        "fontSize": 17,
                      }
                    }
                    testID="listItemTitle"
                  >
                    test2
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </RNCSafeAreaView>
  </Modal>
</View>
`;
